jwst/ami/bp_fix.py

```diff
@@ -0,0 +1,421 @@
+
+import astropy.io.fits as fits
+import numpy as np
+import logging
+import os
+
+from copy import deepcopy
+from poppy import matrixDFT
+from scipy.ndimage import median_filter
+
+from stdatamodels.jwst.datamodels import dqflags
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+
+"""pipeline implementation of Jens Kammerer's bp_fix code based on Ireland 2013 algorithm"""
+
+micron = 1.0e-6
+filts = ['F277W', 'F380M', 'F430M', 'F480M', 'F356W', 'F444W']
+# TO DO: get these from some context-controlled place
+filtwl_d = {  # pivot wavelengths
+    'F277W': 2.776e-6,  # less than Nyquist
+    'F380M': 3.828e-6,
+    'F430M': 4.286e-6,
+    'F480M': 4.817e-6,
+    'F356W': 3.595e-6,  # semi-forbidden
+    'F444W': 4.435e-6,  # semi-forbidden
+}
+filthp_d = {  # half power limits
+    'F277W': (2.413e-6, 3.142e-6),
+    'F380M': (3.726e-6, 3.931e-6),
+    'F430M': (4.182e-6, 4.395e-6),
+    'F480M': (4.669e-6, 4.971e-6),
+    'F356W': (3.141e-6, 4.068e-6),
+    'F444W': (3.880e-6, 5.023e-6),
+}
+WL_OVERSIZEFACTOR = 0.1  # increase filter wl support by this amount to 'oversize' in wl space
+
+# GET PUPIL MASK FROM WEBBPSF
+webbpsf_path = os.getenv('WEBBPSF_PATH')
+pupilfile_nrm = os.path.join(webbpsf_path,'NIRISS/optics/MASK_NRM.fits.gz')
+nrm_pupil = fits.getdata(pupilfile_nrm)
+
+#clearp_pupil = fits.getdata(pupilfile_clearp)
+pupil_masks = {
+    "NRM": nrm_pupil,
+#    "CLEARP": clearp_pupil,
+}
```

Comment:
```suggestion
pupil_masks = {}
webbpsf_path = os.getenv('WEBBPSF_PATH')
if webbpsf_path is not None:
    pupilfile_nrm = os.path.join(webbpsf_path,'NIRISS/optics/MASK_NRM.fits.gz')
    pupil_masks["NRM"] = fits.getdata(pupilfile_nrm)
else:
    # TODO maybe a warning here that WEBBPSF_PATH is not defined?
    # warnings.warn("WEBBPSF_PATH environment variable is undefined...")
```

I wonder if something like this might fix the docs build. This would leave `pupil_masks["NRM"]` undefined when `WEBBPSF_PATH` is not defined so some of the other code in this PR will almost certainly fail but I'm not sure if the docs require any of that to run.

Reply:
I can make that change so hopefully the docs build would work, though I think it would have the same issue when it encounters the similar lines in ami/utils.py: https://github.com/rcooper295/jwst/blob/amianalyze/jwst/ami/utils.py#L1220
I could also have it raise an exception in the beginning ami_analyze_step.py where I check some of the arguments if WEBBPSF_PATH isn't in the environment, since the code will break without it?

Reply:
The docs look almost built :) The issue now appears to be a missing `throughput_reffile` file which is referenced in `docs/jwst/references_general/references_general.rst`. Should there be a `throughput_reffile` in the folder with that file?

Reply:
Getting there, one error at a time... No, I removed the use of the throughput ref file as part of the changes to the step (it used too few wavelengths across the bandpass to be useful). I'll remove the reference to the throughput file in the docs there and see if that fixes it. I have a few more documentation changes to make still that hopefully won't break it again.

Reply:
Unfortunately I know nothing about that file or if it should be removed from crds.

The good news is... the docs built!